src/main/java/com/exactpro/epfast/decoder/message/FastCompiler.java

```diff
@@ -0,0 +1,111 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+package com.exactpro.epfast.decoder.message;
+
+import com.exactpro.epfast.decoder.message.commands.*;
+import com.exactpro.epfast.decoder.message.commands.InitIndexedProperty;
+import com.exactpro.epfast.template.*;
+
+import java.util.*;
+
+public class FastCompiler {
+
+    private final ArrayList<DecoderCommand> commandSet = new ArrayList<>();
+
+    private FastCompiler() {
+    }
+
+    public static Map<Reference, List<DecoderCommand>> compile(Collection<? extends Template> templates) {
+        HashMap<Reference, List<DecoderCommand>> commandSets = new HashMap<>();
+        for (Template template : templates) {
+            commandSets.put(
+                template.getTemplateId(),
+                compileSubroutine(template.getTypeRef(), template.getInstructions()));
+        }
+        return commandSets;
+    }
+
+    private static ArrayList<DecoderCommand> compileSubroutine(
+            Reference typeRef, Collection<? extends Instruction> instructions) {
+        FastCompiler compiler = new FastCompiler();
+        compiler.compile(typeRef, instructions);
+        return compiler.commandSet;
+    }
+
+    private void compile(
+            Reference typeRef, Collection<? extends Instruction> instructions) {
+        if (typeRef != null) {
+            commandSet.add(new InitApplicationType(typeRef));
+        }
+        for (Instruction instruction : instructions) {
+            if (instruction instanceof Group) {
+                compileGroup((Group) instruction);
+            } else if (instruction instanceof Sequence) {
+                compileSequence((Sequence) instruction);
+            } else if (instruction instanceof FieldInstruction) {
+                compileFieldInstruction((FieldInstruction) instruction);
+            } else if (instruction instanceof TemplateRef) {
+                TemplateRef templateRef = (TemplateRef) instruction;
+                // TODO dereference
+                commandSet.add(new CallByReference(templateRef.getTemplateRef()));
+            }
+        }
+        commandSet.add(new RetApplicationType());
+    }
+
+    private void compileGroup(Group group) {
+        commandSet.add(new StaticCall(compileSubroutine(group.getTypeRef(), group.getInstructions())));
+        commandSet.add(new SetApplicationTypeProperty(group.getFieldId()));
+    }
+
+    private void compileSequence(Sequence sequence) {
+        if (sequence.isOptional()) {
+            commandSet.add(new ReadNullableInt32());
+            commandSet.add(new SetNullableLengthField());
+        } else {
+            commandSet.add(new ReadMandatoryInt32());
+            commandSet.add(new SetMandatoryLengthField());
+        }
+        commandSet.add(new InitIndexedProperty(sequence.getFieldId()));
```

Comment:
Due to Java limitation we can't actually use whole uin32 value space for indexed values. Add check in `InitIndexedProperty` for value more than INT_MAX